src/ekka_rlog_agent.erl

```diff
@@ -0,0 +1,159 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc This module implements a gen_statem which pushes rlogs to
+%% a remote node.
+%%
+%% The state machine consists of 3 states:
+%% `catchup', `switchover', `normal', which are explained in detail below:
+%%
+%%
+%% All sends are done as `gen_rpc' calls to the replicant node.
+
+-module(ekka_rlog_agent).
+
+-behaviour(gen_statem).
+
+%% API:
+-export([start_link/3, stop/1]).
+
+%% gen_statem callbacks:
+-export([init/1, terminate/3, code_change/4, callback_mode/0, handle_event/4]).
+
+-include("ekka_rlog.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%% Define macros for each state to prevent typos:
+-define(catchup, catchup).
+-define(switchover, switchover).
+-define(normal, normal).
+
+-type state() :: ?catchup | ?switchover | ?normal.
+
+-record(d,
```

Comment:
we'll have to deal with hot beam upgrades.
extending record fields is always painful and has to be synced.
maybe use map ?

Reply:
My guess is that if the fields change in the hot upgrade, then something is changed in the transaction replication mechanism. This is painful by itself. Perhaps in most cases we'll have to restart agents and re-sync the replicants.

Reply:
it could be just a flag or something like that to be added, which does not impact the replication protocol.